OPA filter implementation && doc reference (#856)
megaease/easegress: doc/reference/filters.md

* **X-Access-Token**: The AccessToken returned by OpenId Connect or OAuth2.0 flow.

## OPAFilter

> The [Open Policy Agent (OPA)](https://www.openpolicyagent.org/docs/latest/) is an open source, 
> general-purpose policy engine that unifies policy enforcement across the stack. It provides a 
> high-level declarative language, we can specify policy as code to enforce policies in 
> Easegress API Gateway. There are 185 builtin operators including functions we can use 
> like examples `net.cidr_contains` and `contains`.

Reviewer: ```suggestion
The [Open Policy Agent (OPA)](https://www.openpolicyagent.org/docs/latest/) is an open-source, 
general-purpose policy engine that unifies policy enforcement across the stack. It provides a 
high-level declarative language, which can be used to define and enforce policies in
Easegress API Gateway. There are 185 built-in operators and functions we can use, 
for examples `net.cidr_contains` and `contains`.
```

```yaml
name: demo-pipeline
kind: Pipeline
flow:
  - filter: opa-filter
    jumpIf: { opaDenied: END }
filters:
  - name: opa-filter
    kind: OPAFilter
    defaultStatus: 403
    readBody: true
    includedHeaders: a,b,c
    policy: |
      package http
      default allow = false
      allow {
         input.request.method == "POST"
         input.request.scheme == "https"
         contains(input.request.path, "/")               
         net.cidr_contains("127.0.0.0/24",input.request.realIP)          
      }
```

> The following table input request fields can be used in OPA policy to help make policy decision enforcement

Reviewer: ```suggestion
The following table lists input request fields that can be used in an OPA policy to help enforce it.
```

| Name                     | Type   | Description                                                           | Example                              |
|--------------------------|--------|-----------------------------------------------------------------------|--------------------------------------|
| input.request.method     | string | The current http request method                                       | "POST"                               |
| input.request.path       | string | The current http request URL path                                     | "/a/b/c"                             |

> | input.request.path_parts | array  | The current http request URL path parts                               | ["a","b","c"]                        |

Reviewer: propose using `segment` instead of `part`, please also update the code.

| input.request.raw_query  | string | The current http request raw query                                    | "a=1&b=2&c=3"                        |
| input.request.query      | map    | The current http request query map                                    | {"a":1,"b":2,"c":3}                  |

> | input.request.headers    | map    | The current http request header map targeted by<br/> includedHeaders  | {"Content-Type":"application/json"}  |

Reviewer: typo, `<br/>`

| input.request.scheme     | string | The current http request scheme                                       | "https"                              | 
| input.request.realIP     | string | The current http request client real IP                               | "127.0.0.1"                          |
| input.request.body       | string | The current http request body string data                             | {"data":"xxx"}                       |

### Configuration

| Name             | Type   | Description                                                                                                            | Required |
|------------------|--------|------------------------------------------------------------------------------------------------------------------------|----------|

> | defaultStatus    | int    | The default http status code when request is denied by OPA policy decision                                             | No       |

Reviewer: ```suggestion
| defaultStatus    | int    | The default HTTP status code when the request is denied by the OPA policy                                             | No       |
```

> | readBody         | bool   | Whether to read request body as OPA policy data on condition than body can be cast to string                           | No       |

Reviewer: ```suggestion
| readBody         | bool   | Whether to read request body as OPA policy data                           | No       |
```

> | includedHeaders  | string | Used to configure which HTTP headers should be included in `input.request.headers`, <br/>comma-separated header names  | No       |

Reviewer: ```suggestion
| includedHeaders  | string | Names of the HTTP headers to be included in `input.request.headers`, comma-separated  | No       |
```

> | policy           | string | The OPA policy written using Rego declarative language                                                                 | Yes      |

Reviewer: ```suggestion
| policy           | string | The OPA policy written in the Rego declarative language                                                                 | Yes      |
```
